### PR TITLE
Add interface to expose out UIView methods

### LINKED ISSUE #13245: Loaded and Unloaded events are not fired correctly
### Description

It seems the `Loaded` and `Unloaded` events of the `VisualElement` class are not always fired correctly.

Here is my test setup:

```xml
<ContentView x:Name="contentView">
    <Label Text="Content" Loaded="OnContentLoaded" Unloaded="OnContentUnloaded" />
</ContentView>
```

The `ContentView` here is for testing purposes to trigger the `Loaded` and `Unloaded` events of the `Label`. Setting the `ControlTemplate` of the `ContentView` would cause the content to be recreated, thus triggering the `Loaded` and `Unloaded` events. Surprisingly, the `Unloaded` and `Loaded` events are not necessarily paired: reloading the `Label` sometimes fires just the `Unloaded` event, without the corresponding `Loaded` event and vice versa. This happens only the first few times when reloading the content. Subsequent reloads would fire the `Unloaded` and `Loaded` events in pairs as expected.

### Steps to Reproduce

1. Run the sample application from the provided repository.
2. Observe the output on the screen showing a single `Loaded` event.
3. Click the `First Template` button to set the first custom `ControlTemplate` to the `ContentView`.
4. Observe the output on the screen showing an `Unloaded` event without a corresponding `Loaded` event.
5. Click the `Second Template` button to set the second custom `ControlTemplate` to the `ContentView`.
6. Observe the output on the screen showing a `Loaded` event without a corresponding `Unloaded` event.
7. Click the `First Template` again button to set the first custom `ControlTemplate` to the `ContentView`.
8. Observe the output on the screen showing both an `Unloaded` event paired with a corresponding `Loaded` event.

### Link to public reproduction project repository

https://github.com/ivan-todorov-progress/maui-loaded-unloaded-event-bug

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS

### Affected platform versions

N/A

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #16860: [regression/8.0.0-preview.7.8842] iOS: Shell navigation crashes the app with SIGSEGV
### Description

The recent update to 8.0.0-preview.7.8842 produces a fatal error in the Shell, where navigation between Flyout items will cause the application to crash.
Sometimes the navigation can be triggered 5-6 times before the crash, sometimes it happens right on the second navigation.
I could reliably reproduce it in DEBUG & RELEASE Mode on an iPad with iOS 16.6.0

Setting the Tag <MauiVersion> in the csproj back to 8.0.0-preview.6.8686 fixes the crash.

This is a video of the app being deployed/started through Visual Studio for Mac with Preview 6. No matter how much I navigate, the app won't crash:

https://github.com/dotnet/maui/assets/67259411/c4b85175-5124-4232-9403-49886728360f

As soon as I switch to Preview 7, the app crashes after a few clicks:

https://github.com/dotnet/maui/assets/67259411/4c6d9465-78e8-4f53-a627-10eefd76af70

It baffles me how this one could slip through. Literally just clicking through the **most advertised** navigation control for the framework crashes the app.... simply astonishing, but not in a good way 😕 

### Steps to Reproduce

1. Open the linked project in Visual Studio for Mac. Make sure you followed the instructions to set it up for .Net 8 development
2. Deploy the linked project onto an iOS Device (Current **stable** version 16.6.0 recommended)
3. Click through the 3 flyout items at the side several times
4. Enjoy the SIGSEGV crash
5. Modify the <MauiVersion> tag in the csproj file and let it point to Maui 8 Preview 6 now
6. Redeploy onto the iOS device
7. Click through the 3 Flyout items relentlessly
8. No crash 

### Link to public reproduction project repository

https://github.com/DDHSchmidt/ShellCrap

### Version with bug

8.0.0-preview.7.8842

### Is this a regression from previous behavior?

Yes, this used to work in .NET MAUI

### Last version that worked well

8.0.0-preview.6.8686

### Affected platforms

iOS

### Affected platform versions

iOS 16.6.0

### Did you find any workaround?

Revert to Maui 8 Preview 6

### Relevant log output

```shell
=================================================================
	Native Crash Reporting
=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
	Native stacktrace:
=================================================================
	0x10452f53c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x10451a528 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044fca34 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x2291c32f4 - /usr/lib/system/libsystem_platform.dylib : <redacted>
	0x1d8e1eb2c - /usr/lib/libobjc.A.dylib : class_getMethodImplementation
	0x1d8e1eb2c - /usr/lib/libobjc.A.dylib : class_getMethodImplementation
	0x1d9f6a79c - /System/Library/Frameworks/Foundation.framework/Foundation : <redacted>
	0x1d9edae24 - /System/Library/Frameworks/Foundation.framework/Foundation : <redacted>
	0x1d9f0caa8 - /System/Library/Frameworks/Foundation.framework/Foundation : <redacted>
	0x1e0f751e0 - /System/Library/Frameworks/QuartzCore.framework/QuartzCore : <redacted>
	0x1e0fcc814 - /System/Library/Frameworks/QuartzCore.framework/QuartzCore : <redacted>
	0x1e1aeec14 - /System/Library/PrivateFrameworks/UIKitCore.framework/UIKitCore : <redacted>
	0x10420f874 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : xamarin_get_runtime_arch
	0x10420fa24 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : xamarin_get_runtime_arch
	0x104574d8c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x10420b888 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : xamarin_release_managed_ref
	0x1041d995c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : 
	0x10453e528 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104532f7c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104530b54 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044ffb68 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044a7a70 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044ab6d0 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104613780 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x10461359c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1d9f08ff8 - /System/Library/Frameworks/Foundation.framework/Foundation : <redacted>
	0x1dfb99154 - /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation : <redacted>
	0x1dfba4dc8 - /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation : <redacted>
	0x1dfb3009c - /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation : <redacted>
	0x1dfb451b8 - /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation : <redacted>
	0x1dfb49da0 - /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation : CFRunLoopRunSpecific
	0x2166c8998 - /System/Library/PrivateFrameworks/GraphicsServices.framework/GraphicsServices : GSEventRunModal
	0x1e1ddb80c - /System/Library/PrivateFrameworks/UIKitCore.framework/UIKitCore : <redacted>
	0x1e1ddb484 - /System/Library/PrivateFrameworks/UIKitCore.framework/UIKitCore : UIApplicationMain
	0x1041f05f4 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : xamarin_UIApplicationMain
	0x10453f610 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x10453e19c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104532ea4 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104530b54 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044ffb68 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044a7a70 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1044adc48 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x104505020 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x10421d00c - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : xamarin_get_original_working_directory_path
	0x10456d638 - /private/var/containers/Bundle/Application/6DE8FB6D-40C5-40C3-BB02-EFEFFD8137E3/ShellCrap.app/ShellCrap : AppleCryptoNative_X509ImportCollection
	0x1fd2e8344 - /usr/lib/dyld : <redacted>

=================================================================
	Basic Fault Address Reporting
=================================================================
Memory around native instruction pointer (0x1d8e25040):0x1d8e25030  f3 03 02 aa f5 03 01 aa f6 03 00 aa 74 00 1e 32  ............t..2
0x1d8e25040  48 10 40 f9 08 8d 7d 92 08 01 40 b9 c8 00 00 37  H.@...}...@....7
0x1d8e25050  68 02 40 f9 08 81 7d 92 08 11 40 f9 08 8d 7d 92  h.@...}...@...}.
0x1d8e25060  08 01 40 b9 48 03 e8 36 e0 03 13 aa e1 03 15 aa  ..@.H..6........

=================================================================
	Managed Stacktrace:
=================================================================
	  at <unknown> <0xffffffff>
	  at Foundation.NSObject:xamarin_release_managed_ref <0x000c2>
	  at Foundation.NSObject:ReleaseManagedRef <0x000e2>
	  at NSObject_Disposer:Drain <0x00170>
	  at <Module>:runtime_invoke_direct_void_NSObject <0x0008e>
	  at <unknown> <0x00000>
	  at <unknown> <0xffffffff>
	  at UIKit.UIApplication:xamarin_UIApplicationMain <0x000c6>
	  at UIKit.UIApplication:UIApplicationMain <0x0006a>
	  at UIKit.UIApplication:Main <0x00158>
	  at ShellCrap.Program:Main <0x0002c>
	  at <Module>:runtime_invoke_direct_void_string[] <0x0007e>
	  at <unknown> <0x00000>
=================================================================
```

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

Currently we're using KO's to monitor changes on iOS UIViews to trigger loaded/unloaded. This is very imprecise but also appears to cause occasional [crashes](https://github.com/xamarin/xamarin-macios/issues/18768). This PR begins the process of moving us to a more precise mechanism for monitoring when views are added/removed from the platform window. 

### Issues Fixed

Fixes #16860
Fixes #13245

### CODE REVIEW

File: src/Core/src/Platform/iOS/MauiTextField.cs [hunk 1]
```diff
@@ -18,10 +19,18 @@ public MauiTextField()
...
+			add => _movedToWindow += value;
+			remove => _movedToWindow -= value;
+		}
+
 		public override void WillMoveToWindow(UIWindow? window)
 		{
 			base.WillMoveToWindow(window);
-			ResignFirstResponderTouchGestureRecognizer.Update(this, window);
+			_movedToWindow?.Invoke(this, EventArgs.Empty);
 		}
```
Comment (mattleibow) [inline src/Core/src/Platform/iOS/MauiTextField.cs:34]: Is this correct? The WillMove uses the Moved event? Maybe we should just have both on the interface?

File: src/Core/src/Platform/iOS/ViewExtensions.cs [hunk 2]
```diff
@@ -650,22 +675,46 @@ internal static IDisposable OnUnloaded(this UIView uiView, Action action)
...
+
+				void OnLifeCycleEventsMovedToWindow(object? sender, EventArgs e)
+				{
+					UnLoadedCheck();
+				}
+			}
+			else
+			{
+				var boundKey = new NSString("bounds");
+				var frameKey = new NSString("frame");
```
Comment (mattleibow) [inline src/Core/src/Platform/iOS/ViewExtensions.cs:710]: How does frame and bounds change when a view is unloaded?

**Review by mattleibow — COMMENTED** [review]
Seems valid and all that. Just had a Q on how the builds works when unloaded and how it can be used to detect unloaded. And I spotted a WillMoveToWindow calling MovedToWindow. Not sure if that should have been 2 events?

File: src/Core/src/Platform/iOS/MauiTextField.cs (hunk 1 above)
Comment (PureWeen) [reply src/Core/src/Platform/iOS/MauiTextField.cs]: fixed

File: src/Core/src/Platform/iOS/ViewExtensions.cs [hunk 3]
```diff
@@ -650,22 +675,36 @@ internal static IDisposable OnUnloaded(this UIView uiView, Action action)
...
-					observers.Remove(observer.Key);
+					observer.Dispose();
 				}
+				observers.Clear();
 			});
 
 			// Ideally we could wire into UIView.MovedToWindow but there's no way to do that without just inheriting from every single
-			// UIView. So we just make our best attempt by observering some properties that are going to fire once UIView is attached to a window.	
-			observers.Add(new NSString("bounds"), (NSObject)uiView.Layer.AddObserver("bounds", Foundation.NSKeyValueObservingOptions.OldNew, (_) => UnLoadedCheck()));
-			observers.Add(new NSString("frame"), (NSObject)uiView.Layer.AddObserver("frame", Foundation.NSKeyValueObservingOptions.OldNew, (_) => UnLoadedCheck()));
```
Comment (mattleibow) [inline src/Core/src/Platform/iOS/ViewExtensions.cs:668]: With the MovedToWindow you originally had the unloaded also listening to the bounds/frame, and now i see you removed it. Did it not work? Is there a case where we are now going to not get an unloaded?

Comment (PureWeen) [reply src/Core/src/Platform/iOS/ViewExtensions.cs:668]: AFAICT it did not work. There's an issue about it here as well https://github.com/dotnet/maui/pull/16868 . 

I ran all of our UITests around loaded/unloaded and it was never triggered by the observers. It was only triggered by the logic that fires unloaded when an element is removed from the logical tree. 

> Is there a case where we are now going to not get an unloaded?

There might be some still, but I'd say there are less after this PR.

**Review by mattleibow — DISMISSED** [review]

**Review by rmarinho — CHANGES_REQUESTED** [review]
`ExcludedWithChildren` device test on iOS failing

File: src/Core/src/Handlers/Button/ButtonHandler.iOS.cs [hunk 4]
```diff
@@ -12,6 +12,10 @@ public partial class ButtonHandler : ViewHandler<IButton, UIButton>
 		// This appears to be the padding that Xcode has when "Default" content insets are used
 		public readonly static Thickness DefaultPadding = new Thickness(12, 7);
 
+		// Because we can't inherit from Button we use the container to handle
+		// Life cycle events and things like monitoring focus changed
+		public override bool NeedsContainer => true;
```
Comment (rmarinho) [inline src/Core/src/Handlers/Button/ButtonHandler.iOS.cs:17]: Isn't this a big change wrapping every button now on a UIView, border, shadows etc?!

File: src/Core/src/Platform/iOS/IUIViewLifeCycleEvents.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Microsoft.Maui.Platform
+{
+	internal interface IUIViewLifeCycleEvents
+	{
+		const string UnconditionalSuppressMessage = 
+			"This event isn't meant to cleanup by itself. The subscriber needs to unsubscribe when it knows it no longer cares about this event.";
```
Comment (rmarinho) [inline src/Core/src/Platform/iOS/IUIViewLifeCycleEvents.cs:8]: We need to start tracking things like this to be translated.

File: src/Core/src/Platform/iOS/MauiImageView.cs [hunk 6]
```diff
@@ -22,12 +23,21 @@ public MauiImageView(CGRect frame)
 		{
 		}
 
+		[UnconditionalSuppressMessage(IUIViewLifeCycleEvents.UnconditionalSuppressMessage, "MA0002")]
+		EventHandler? _movedToWindow;
```
Comment (rmarinho) [inline src/Core/src/Platform/iOS/MauiImageView.cs:27]: Can we add a explanation why supressing?

File: src/Core/src/Handlers/Button/ButtonHandler.iOS.cs (hunk 4 above)
Comment (PureWeen) [reply src/Core/src/Handlers/Button/ButtonHandler.iOS.cs:17]: Ish? Containers in theory shouldn't be a big change. They should be mostly unnoticeable from a behavior stand point. This is iOS as well so not as worried about performance

File: src/Core/src/Platform/iOS/MauiImageView.cs (hunk 6 above)
Comment (PureWeen) [reply src/Core/src/Platform/iOS/MauiImageView.cs:27]: Do we need an additional one in addition to the const? 

https://github.com/dotnet/maui/pull/16868/files#diff-f5450e00e8422d29dd1d2671fc035691ad5bdd53d086bd9b08227f708d8fa3cdR7-R8

Comment (rmarinho) [reply src/Core/src/Platform/iOS/MauiImageView.cs:27]: I thinks we are using justification as category , maybe we want something like : https://github.com/dotnet/maui/blob/01fc5eb0df4a1e5acbb947de313966e533f9ce09/src/Core/src/Platform/iOS/MauiPicker.cs#L22
```
		[UnconditionalSuppressMessage("Memory", "MA0002", Justification = IUIViewLifeCycleEvents.UnconditionalSuppressMessage)]

```

Comment (PureWeen) [reply src/Core/src/Platform/iOS/MauiImageView.cs:27]: Fixed

**Review by rmarinho — APPROVED** [review]

**Pull request merged**